### PR TITLE
Protect for table lookup zero divisor

### LINKED ISSUE #10464: Fatal error in heating and cooling loads during sizing run
Issue overview
--------------
This error is coming up on some models in the openstudio-standards CI with the latest develop branch of OpenStudio and EnergyPlus 24.1.
```
[openstudio.model.Model] The run did not finish and had following errors: DualSetPointWithDeadBand: Unanticipated combination of heating and cooling loads - report to EnergyPlus Development Team  occurs in Zone=F2 N1 APARTMENT ZN During Warmup, Environment=RUN PERIOD 1, at Simulation time=01/01 00:00 - 00:10  LoadToHeatingSetPoint=-N0AN, LoadToCoolingSetPoint=-N0AN  Zone Heating Set-point=21.70  Zone Cooling Set-point=24.40  Zone TempDepZnLd=-N0AN  Zone TempIndZnLd=-N0AN  Zone ThermostatSetPoint=21.70\
Program terminates due to above conditions.
```
 The tests are passing in OS v3.7.0.

.idf and .epw file to replicate:
[error.zip](https://github.com/NREL/EnergyPlus/files/14951657/error.zip)

To narrow things down, I made a copy of the model in E+ 23.2 and then forward translated it, which runs fine.
The difference is in the new curve name fields attached to the HeatExchanger:AirToAir:SensibleAndLatent object. The default curve tables from OpenStudio seem to be causing the issue. Related OpenStudio PR: https://github.com/NREL/OpenStudio/pull/5099.

### Details
Some additional details for this issue:
 - Windows
 - OpenStudio develop (pre v3.8 release)
 - E+ 24.1

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by Myoldmopar**
Pull request overview
---------------------
 - Fixes #10464 

The actual problem wasn't in the load calculations, it was a NaN being propagated after a divide by zero.  In the table lookup objects. users can enter a normalization divisor value, and a zero value was allowed to pass through without being checked.  This resulted in a NaN value coming out of the curve evaluation, which was immediately causing problems in other parts of the code.

I added a trap for zero value on the normalization divisor and we now issue a severe error.  I added a unit test that ensures the zero value is caught properly.

### CODE REVIEW

File: src/EnergyPlus/CurveManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/CurveManager.cc:2501]: Check if the value coming in is zero (or essentially zero).  This is only hit if the normalization divisor is going to be used (based on other input fields).  I'm open to wording improvement, but it should get the job done.

File: tst/EnergyPlus/unit/CurveManager.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/CurveManager.unit.cc:519]: New unit test that sets up a simple table lookup with zero valued normalization divisor.  It first makes sure the runtime error is thrown during curve get input, and then also makes sure there is a specific mention of the normalization divisor in the error output.  Note I'm not comparing the whole error output because I don't care about overall output message structure, I just want to make sure the divisor error is, in fact, emitted.

File: tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:212]: Added a new worker function on the unit test fixture.  This allows you to compare some specific search phrase in the ERR contents without having to list the entire formatted ERR message.  See the unit test above in this PR for an example.

File: tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.hh
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.hh:205]: New worker function in fixture header so it can be used by unit tests.

**Review by Myoldmopar — COMMENTED** [review]
OK, I think this is good to go if CI agrees.  If anyone wants to review, that's welcomed, but it's pretty minimal.

**Review by nealkruis — APPROVED** [review]

File: src/EnergyPlus/CurveManager.cc
Comment (jmarrec) [inline src/EnergyPlus/CurveManager.cc]: std::abs, or if you insist std::fabs
But std:: namespace it.
@myoldmopar

Comment (jmarrec) [reply src/EnergyPlus/CurveManager.cc]: In c++ you have function overloading, so std::abs is sufficient.

In C, abs takes an int, fabs a double, which is a common cause of problems.

**Comment by Myoldmopar** [comment]
There you go @jmarrec, thanks!  OK, if this is clean again after lunch I'll merge.

**Comment by Myoldmopar** [comment]
All good, merging.

File: tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc
Comment (jmarrec) [inline tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:208]: I think you meant search_string instead of harcoding?

Comment (Myoldmopar) [reply tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:208]: No I wanted it to search for this hard coded string.


. 

.


.
...





OK, OK, I'll slow down.  Thank you an awful lot, @jmarrec !  Fix incoming soon - tomorrow.

Comment (jmarrec) [reply tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:208]: Reading only the first bit on my phone in the email notification cracked me up ;)

Comment (jmarrec) [reply tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:208]: Clang-tidy 15 definitely catches it for me. I wonder why the cppcheck didn't? And that's because we only run cppcheck in `./src` :)

**Pull request merged**
